### DISCUSSION TITLE
2021.10.06 멘토링

### DISCUSSION THREAD

**Discussion by kwakjeeyoon** [boostcampaitech2/object-detection-level2-cv-06#89] [General]
**지우님** : mmdetection 익숙해지고 augmentation 관련 내용을 살펴보고 있다.

- mmdetection 으로 하나로 통일해야 협업을 할수 있는것 같은데 다른 프레임워크를 쓰면 실험을 어떻게 할지 → 베이스라인이 정해져 있지 않기 때문에 어려운 문제인듯 하다.
- 좋은 템플릿을 하나 찾아 쓰는 것도 좋다. 베이스라인 말고도 더 좋게 만들어진 템플릿으로 협업을 하는 것도 좋을 것 같다.
- object detection 과정 자체가 매우 복잡해 모듈화하기 어려워 mmdetection 을 써야 겠다고 생각했다.

**지윤님** : model sota 코드를 보다가 못하겠어서 kaggle competition을 보면서 model 살펴보고 있었다. 

- 모델 선정 : 문제상황 정의 - 제약상황이 많은 경우가 있다. 제약상황에 맞게 찾아보고 진짜 대충 5분 이내로 모델을 찾아보고 논문을 읽은 다음 완성된 코드가 있는 모델을 찾는다. + pretrained model이 있는 모델. 우리의 computing resource
- Yolov5 많이 쓴다.  - 성능도 괜찮고 속도도 빠르다.
- competitoin의 경우 0.1 성능이 더 좋은것을 선택하지만 하이퍼파라미터 튜닝도 공부에 도움이 안된다고 생각한다. 다양한 실험을 많이 해보는게 공부에 도움이 되는것 같다.
- paper with code 많이 참고 하고 레퍼런스 많이 되는 모델 위주로 찾거나 아니면 vote 수가 많은 모델을 쓴다. task에 있어서 자주 쓰이는 모델들이 있다. 하지만 model select 보다 데이터셋이나  다른 기술을 더 써보는게 더 중요

**지유님** : config에 내용을 추가할때마다 error가 생겨 해결하는데 시간을 보냈다. github 협업에 대해 많이 배웠다. 

- 배경을 제거해서 합성하는 등의 방법은 어떨까 → 좋은 것 같다. 데이터셋이 많이 없을때 사용하면 좋다. 합성을 해서 성능이 무조건 좋아지는 것이 아니라 내가 가지고 있는 데이터셋과 리얼 데이터셋이 비슷하게 될 수 있도록 처리하는 것은 좋으나 무조건 합성을 한다고 좋아지지는 않는다.
    
    ⇒ 거의 바이블일 정도로 많이 하는 방법이다. 
    

**나은님** : object detection 공부를 처음 해봐서 segmentation 모델을 보다가 실패하였다.

**재환님** : 협업을 하기 위해 mmdetection 을 베이스라인을 잡았지만 custom을 하는 과정에서 어려움이 많았다. github 협업 과정이 주된 공부 내용이었다. 실험을 위한 세팅을 하는 단계. 

**준혁님** : 협업 관련해서 github wandb 를 처음써서 어려움이 있었다. 모델이 좋은 이유를 아는 것이 힘들었다. 

- 작은 bbox의 경우는 edge case라고 볼 수 있나? - 가치 판단의 영역인것 같은데 대응을 해야하는 크기라면 이미지 크기 줄이고 늘리면서 반응할 수 있다. 최대한 클래스의 분포가 continuous 하게 유지되게 하면 딥러닝의 학습이 더 잘 될 수 있다. (1px정도면 대응할 필요가 없다. - 버린다 마인드, 넣는다고 학습이 될까?)
- bounding box를 늘려서 box가 잘리는 경우 같은 상황이 발생하지 않도록 신경써서 잘 사용하는 것이 중요하다.

**경재님** : 세팅을 빨리 끝내고 실험단계로 넘어가야 된다고 생각을 했지만 생각대로 안되었고 이제야 세팅이 끝나는 단계라서 아쉬웠고 리액션이 많이 없어서 섭섭했다. 대화가 많아지면 좋겠다. 

- docker를 써보는 것도 방법이 될수 있다. 도커 이미지를 돌리면 모두가 똑같은 환경에서 작업을 할 수 있다. 경로도 데이터셋, 모든 파일의 경로가 똑같기 때문에 모두가 똑같은 환경에서 작업할 수 있다.
- 만들면 어느정도 용량? - 데이터는 링크만 넣으면 경로가 달라질수도 있어서 데이터까지 도커에 넣어서 사용하면 좋다.
- 도커 많이 쓰기 때문에 써보면 좋다.
- SQL 보다 도커가 백배 더 중요하다. 아니다 한... 세네배? 중요하다.
- 5000장 정도면 적당한 양인것 같다.

---

- 개발자의 협업에는 슬랙과 github이 팔할이다.
- 분위기가 좋은 팀이 성적이 좋다.
- mmdetection 커스텀이 어렵고 협업의 경험이 없는 사람은 따라가기가 어렵고 협업 세팅에 시간이 많이 결렸다는 것이 주된 내용인것 같다.
- 빠르게 손절하는 것도 답이다. 좋은 것을 선택하는 것이 좋고 장단점을 파악하고 단점이 크면 다른 방법으로 해라.
- 모듈화로 mmdetection, 코드 등 호환이 되게 코드 템플릿을 만들고 ood 에 맞게 짜던가 등등의 방법이 있다. 어려운 task이기 때문에 생각해봐라.
- 코드 리뷰를 많이 하는 것이 좋다.
- 강의 템포가 빠른것은 짧은 기간동안 딥러닝의 모든것을 커버하기 위해 압축이 많이 되어 그런것. 이론에서 이해를 하겠다는 것은 불가능에 가깝다. 모든걸 이해하면 천재이다.
    
    ⇒ 추천하는 방향은 이론에서 이런게 있다더라 라는 정도만 보고 실전에서 부딪히면서 공부하는게 빠르게 배우는 부캠에 맞는 방법인 것 같다. (top-down 방식)
    
    멘토들도 모르는 내용이 있다. 기죽지 마라. 이론은 시간의 문제이기 때문에 진짜 필요한 능력은 다양한 모델을 사용해봐야 는다. 
    
- 커뮤니케이션 능력이 중요하다.
- 문제해결능력을 기르기 위해 EDA를 통해 문제 해결의 가설을 세우고 아이디어를 내보고 구현을 해보고 하고 커뮤니케이션을 하는 등등의 경험을 많이 해보는 것이 중요하다.
- mmdetection 은 잘못 만든 프레임워크 이지만 앞으로 개같은 코드를 만질 일이 생긴다. 이 경험이 도움이 되기 때문에 버리는 시간이라고 생각하지 말고 삽질을 많이 해봐라. 개발을 하다보면 내가 모르는 기술을 쓰는 일이 생기기 때문에 나중에 도움이 많이 된다.
- 자주 쓰는 code temlate 을 저장해서 써봐라.

**Comment by jaehwan-AI** [comment]
> SQL 보다 도커가 백배 더 중요하다. 아니다 한... 세네배? 중요하다.

디테일에 소름돋았습니다 ㅋㅋㅋ
